vortex-serde/src/layouts/read/mod.rs

```diff
@@ -1,108 +1,78 @@
-    ///
-    /// The layout can either return a batch data, i.e. an Array or ask for more layout messages to
-    /// be read. When requesting messages to be read the caller should populate the messages in the cache
-    /// and then call back into this function.
+    /// Register all horizontal boundaries of this layout.
     ///
-    /// The layout is finished reading when it returns None
-    fn read_next(&mut self) -> VortexResult<Option<ReadResult>>;
+    /// Layout should register its begging adjusted by row_offset and do so for all of its children
+    fn add_splits(&self, row_offset: usize, splits: &mut BTreeSet<usize>) -> VortexResult<()>;
 
-    // TODO(robert): Support stats pruning via planning. Requires propagating all the metadata
-    //  to top level and then pushing down the result of it
-    // Try to use metadata of the layout to perform pruning given the passed `Scan` object.
-    //
-    // The layout should perform any planning that's cheap and doesn't require reading the data.
-    // fn plan(&mut self, scan: Scan) -> VortexResult<Option<PlanResult>>;
-}
-
-pub fn null_as_false(array: BoolArray) -> VortexResult<Array> {
-    Ok(match array.validity() {
-        Validity::NonNullable => array.into_array(),
-        Validity::AllValid => {
-            BoolArray::try_new(array.boolean_buffer(), Validity::NonNullable)?.into_array()
-        }
-        Validity::AllInvalid => BoolArray::from(BooleanBuffer::new_unset(array.len())).into_array(),
-        Validity::Array(v) => {
-            let bool_buffer = &array.boolean_buffer() & &v.into_bool()?.boolean_buffer();
-            BoolArray::from(bool_buffer).into_array()
-        }
-    })
-}
-
-#[cfg(test)]
-mod tests {
-    use vortex::array::BoolArray;
-    use vortex::validity::Validity;
-    use vortex::IntoArrayVariant;
-
-    use super::*;
-
-    #[test]
-    fn coerces_nulls() {
-        let bool_array = BoolArray::from_vec(
-            vec![true, true, false, false],
-            Validity::Array(BoolArray::from(vec![true, false, true, false]).into()),
-        );
-        let non_null_array = null_as_false(bool_array).unwrap().into_bool().unwrap();
-        assert_eq!(
-            non_null_array.boolean_buffer().iter().collect::<Vec<_>>(),
-            vec![true, false, false, false]
-        );
-    }
+    /// Reads the data from the underlying layout within given selection
+    ///
+    /// Layout is required to return all data for given selection in one batch.
+    /// Layout can either return a batch data (i.e., an Array) or ask for more layout messages to
+    /// be read. When requesting messages to be read the caller should populate the message cache used
+    /// when creating the invoked instance of this trait and then call back into this function.
+    ///
+    /// The layout is finished producing data for selection when it returns None
+    fn read_selection(&mut self, selector: RowMask) -> VortexResult<Option<BatchRead>>;
```

Comment:
I think this RowMask should be `&RowMask`? Surely it should be immutable? Also you're doing an awful lot of BitMap cloning